File: ios/PortalWidget/Networking/PostDTO.swift
//
//  PostDTO.swift
//  PortalWidgetExtension
//
//  Created by 하정우 on 11/29/25.
//

import Foundation
import Moya

struct ResultDTO: Codable {
  var results: [PostDTO]
}

struct PostDTO: Codable {
  var id: Int
  var title: String
  var araID: Int
  var author: String
  var date: String
  
  enum CodingKeys: String, CodingKey {
    case id
    case title
    case araID = "ara_article"
    case author = "writer_department"
    case date = "registered_at"
  }
}

struct KeywordListDTO: Codable {
  var id: Int
  var title: String
}

extension PostDTO {
  func toModel(reason: DisplayReason) -> Post {
    Post(
      id: self.id,
      title: self.title,
      author: self.author,
      reason: reason,
      date: date.date ?? Date.distantPast
    )
  }
}

extension KeywordListDTO {
  func getPost(keyword: String) async -> Post? {
    let provider = MoyaProvider<PortalWidgetTarget>()
    
    do {
      let response = try await provider.request(.fetchNoticeByPostId(postId: self.id))
      let _ = try response.filterSuccessfulStatusCodes()
      
      let data = try response.map(PostDTO.self).toModel(reason: .keyword(word: keyword))
      return data
    } catch {
      return nil
    }
  }
}

File: android/app/src/main/java/com/example/new_ara_app/networkings/PostDTO.kt
package com.example.new_ara_app.networkings

import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import com.example.new_ara_app.helpers.toDate
import com.example.new_ara_app.models.DisplayReason
import com.example.new_ara_app.models.Post
import com.google.gson.annotations.SerializedName
import java.util.Date

data class PostDTO(
    val title: String,
    @SerializedName("ara_article") val araID: Int,
    @SerializedName("writer_department") val author: String,
    @SerializedName("registered_at") val date: String,
) {
    @RequiresApi(Build.VERSION_CODES.O)
    fun toModel(reason: DisplayReason): Post {
        return Post(
            id = this.araID,
            title = this.title,
            author = this.author,
            reason = reason,
            date = date.toDate() ?: Date(Long.MIN_VALUE)
        )
    }
}

data class KeywordPostDTO(
    val id: Int,
    val title: String,
) {
    @RequiresApi(Build.VERSION_CODES.O)
    suspend fun getPost(keyword: String, api: PortalWidgetService, context: Context): Post? {
        return try {
            val post = api.fetchNoticeByPostId(this.id)
            post.toModel(DisplayReason.Keyword(keyword))
        } catch (e: Exception) {
            null
        }
    }
}

data class KeywordResultDTO(
    val results: List<KeywordPostDTO>
)